### PR TITLE
[Merged by Bors] - feat: Topological properties of order-connected sets in ℝⁿ

### PR DESCRIPTION
**Pull request by YaelDillies**
Partially forward-port https://github.com/leanprover-community/mathlib/pull/16976

---

- [x] depends on: #11606

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Analysis/Normed/Order/UpperLower.lean
Comment (urkud) [inline Mathlib/Analysis/Normed/Order/UpperLower.lean:162]: You should either introduce a typeclass (and prove instances for `Real`, `Prod`, and `∀ i, X i`) or use `pi` somewhere in the names of all these lemmas.

Also, `Bornology` lemmas should assume `Finite`, not `Fintype`.

Comment (YaelDillies) [reply Mathlib/Analysis/Normed/Order/UpperLower.lean:162]: This used to require `Fintype` because this was a lemma about `Metric.IsBounded`

File: Mathlib/Topology/Order/Bounded.lean
Comment (urkud) [inline Mathlib/Topology/Order/Bounded.lean:26]: Why do you delete these lemmas instead of generalizing/deprecating them? I see that they should use `Filter.IsBoundedUnder`, what are other reasons?

File: Mathlib/Analysis/Normed/Order/UpperLower.lean
Comment (urkud) [inline Mathlib/Analysis/Normed/Order/UpperLower.lean:220]: https://leanprover-community.github.io/contribute/style.html#whitespace-and-delimiters says to use `<|`, not `$`.

File: Mathlib/Analysis/Normed/Order/UpperLower.lean
Comment (urkud) [inline Mathlib/Analysis/Normed/Order/UpperLower.lean:213]: Please add `pi` somewhere in the name (here and below). Also please add a TODO about finding a right generality for this lemma.

File: Mathlib/Topology/Order/Bounded.lean
Comment (YaelDillies) [reply Mathlib/Topology/Order/Bounded.lean:26]: They are a simple combination of easier lemmas belonging to different areas, namely `Bornology.IsBounded.bddBelow`/`Bornology.IsBounded.bddAbove` and `isBoundedUnder_of` and some quantifier reshuffle. I did not deprecate them because
* I want to be able to remove the file directly
* I cannot state them using `ℝ` in `Topology.Order.Bornology`
* I don't want to have to think about how I should restate them
* I can't redirect to a single lemma within the deprecation
* They have been added for exactly one use anyway

Writing that message, I have now spent too much time thinking about those two lemmas.

File: Mathlib/Analysis/Normed/Order/UpperLower.lean
Comment (Ruben-VandeVelde) [inline Mathlib/Analysis/Normed/Order/UpperLower.lean:29]: Did you mean "appended"?

File: Mathlib/Analysis/Normed/Order/UpperLower.lean
Comment (eric-wieser) [inline Mathlib/Analysis/Normed/Order/UpperLower.lean:209]: What happened to `is_antichain.interior_eq_empty`?

File: Mathlib/Data/Real/NNReal.lean
Comment (eric-wieser) [inline Mathlib/Data/Real/NNReal.lean:1]: If you split the changes in this file to its own PR, I'll merge it immediately (in the interest of removing items from out-of-sync)

File: Mathlib/Analysis/Normed/Order/UpperLower.lean
Comment (YaelDillies) [reply Mathlib/Analysis/Normed/Order/UpperLower.lean:209]: It has already been ported (and generalised) as `IsAntichain.interior_eq_empty`

File: Mathlib/Data/Real/NNReal.lean
Comment (YaelDillies) [reply Mathlib/Data/Real/NNReal.lean:1]: #11606

File: Mathlib/Topology/Order/Bornology.lean
Comment (Ruben-VandeVelde) [inline Mathlib/Topology/Order/Bornology.lean:78]: I think you might have another conflict with yourself on `inter_subset_left`

File: Mathlib/Topology/Order/Bounded.lean
Comment (Ruben-VandeVelde) [reply Mathlib/Topology/Order/Bounded.lean:26]: I think our current policy would lean towards deprecating them

**Review by Ruben-VandeVelde — APPROVED** [review]
Seems pretty straightforward, modulo the removal of `Topology.Order.Bounded`

File: Mathlib/Topology/Order/Bornology.lean
Comment (YaelDillies) [reply Mathlib/Topology/Order/Bornology.lean:78]: Well spotted

File: Mathlib/Topology/Order/Bounded.lean
Comment (YaelDillies) [reply Mathlib/Topology/Order/Bounded.lean:26]: I have restored `Topology.Order.Bounded`

**Review by Ruben-VandeVelde — APPROVED** [review]
maintainer merge

File: Mathlib/Analysis/Normed/Order/UpperLower.lean
Comment (fpvandoorn) [inline Mathlib/Analysis/Normed/Order/UpperLower.lean:137]: Should these lemmas also get a `_pi` ending?

**Review by fpvandoorn — COMMENTED** [review]
One more comment. Feel free to merge even if you disagree.

bors d+

File: Mathlib/Analysis/Normed/Order/UpperLower.lean
Comment (j-loreaux) [inline Mathlib/Analysis/Normed/Order/UpperLower.lean:31]: This comment shouldn't hold up this PR, but the relevant lemmas could just take a homeomorphism with a pi-type as an explicit argument. It's not a pretty solution, but I think it would work, and I believe we already have the necessary homeomorphisms available.

(On second thought, would it need to be an order isomorphism too? That probably rules out this idea as being useful)

Comment (YaelDillies) [reply Mathlib/Analysis/Normed/Order/UpperLower.lean:31]: Yeah, it would also need to be an order isomorphism. That's the crux. We need a way to say "Order on this type is given by those coordinates", which sounds a lot like the `Inducing`/`Embedding` business in the topology part of the library.

Food for thought, I guess

File: Mathlib/Analysis/Normed/Order/UpperLower.lean
Comment (YaelDillies) [reply Mathlib/Analysis/Normed/Order/UpperLower.lean:137]: Good point

**Comment by YaelDillies** [comment]
bors merge

**Comment by YaelDillies** [comment]
bors merge

**Pull request closed without merging**
